### ISSUE TITLE
Option/configuration/extension for smaller exception tracebacks

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#6712]
---

## Feature request

Following on from today's public meeting: https://github.com/numba/numba/wiki/Minutes_2021_02_09 where an idea along these lines was discussed:

> A configuration or extension point to make Numba do a short "single line error report" might be useful for use cases where a product wraps or consumes Numba and also takes user input as something to attempt to compile. In this case simply conveying to the user that "Numba could not compile line X" may be sufficient for the needs of the project.

As noted at the meeting, I have a strong preference for Numba to continue with its current default of it having compiler-like error messages, enhancements to improve these are welcomed. However, as noted above, having a way to request Numba to act more in a "consumption" mode where tracebacks are squashed sounds like a good fit for the use cases described.

This ticket is to track and discuss.

CC @ehsantn
CC @jpivarski

**Comment by jpivarski** [comment]
I agree that it shouldn't be default. It could be a good argument for decorators:

```python
@nb.jit(errors="simple")
def ...
```

If the current error is structured as an object (i.e. doesn't have to be parsed as a string), then it would be easier to write customizations.

Furthermore, I can ask users to collect examples of error messages, the code that produced them, and what those users wanted the errors to say. Maybe that collection point is this issue thread, or maybe it could be a GitHub template for the "new issue" button (so that the template can remind them to include all three parts: current message, code that produced it, and what they wanted).

**Comment by ehsantn** [comment]
CC @hadia206
CC @juanjgalvez

**Comment by stuartarchibald** [comment]
Thanks @jpivarski 

> I agree that it shouldn't be default. It could be a good argument for decorators:
> 
> ```python
> @nb.jit(errors="simple")
> def ...
> ```
> 
> If the current error is structured as an object (i.e. doesn't have to be parsed as a string), then it would be easier to write customizations.

The errors currently undergo a lot of rewriting, in part because there's often a large number of them that all need gluing together to produce information about all the things that may have gone wrong, all as a single exception. It would be nice if these had more structure such that an error output adapter type thing could be registered.

> Furthermore, I can ask users to collect examples of error messages, the code that produced them, and what those users wanted the errors to say. Maybe that collection point is this issue thread, or maybe it could be a GitHub template for the "new issue" button (so that the template can remind them to include all three parts: current message, code that produced it, and what they wanted).

I think this is a good idea. Collecting a few samples of:
```
I did X:
I expected error:
I got error:
```
May well enable common patterns to be spotted/help guide what needs to be done. Thanks for your help :)
